In [1]:
print("Hello World")


Hello World


## **WALMART SALES DATA CLEANING**

#### **Step 1 DATA EXPLORATION & LEADING**

In [2]:
#importing dependencies
import pandas as pd

#mysql toolkit

import pymysql # this will work as an adapter
from sqlalchemy import create_engine

#psql
#psycopg2  pip install psycopg2

In [3]:
print(pd.__version__)

2.3.2


In [4]:
df = pd.read_csv('Walmart.csv', encoding_errors='ignore')
df.shape

(10051, 11)

In [5]:
df.head()

invoice_id   Branch         City                category unit_price  \
0           1  WALM003  San Antonio       Health and beauty     $74.69   
1           2  WALM048    Harlingen  Electronic accessories     $15.28   
2           3  WALM067  Haltom City      Home and lifestyle     $46.33   
3           4  WALM064      Bedford       Health and beauty     $58.22   
4           5  WALM013       Irving       Sports and travel     $86.31   

   quantity      date      time payment_method  rating  profit_margin  
0       7.0  05/01/19  13:08:00        Ewallet     9.1           0.48  
1       5.0  08/03/19  10:29:00           Cash     9.6           0.48  
2       7.0  03/03/19  13:23:00    Credit card     7.4           0.33  
3       8.0  27/01/19  20:33:00        Ewallet     8.4           0.33  
4       7.0  08/02/19  10:37:00        Ewallet     5.3           0.48

In [6]:
df.tail()

invoice_id   Branch        City             category unit_price  \
10046        9996  WALM056     Rowlett  Fashion accessories        $37   
10047        9997  WALM030  Richardson   Home and lifestyle        $58   
10048        9998  WALM050    Victoria  Fashion accessories        $52   
10049        9999  WALM032       Tyler   Home and lifestyle        $79   
10050       10000  WALM069    Rockwall  Fashion accessories        $62   

       quantity      date      time payment_method  rating  profit_margin  
10046       3.0  03/08/23  10:10:00           Cash     3.0           0.33  
10047       2.0  22/02/21  14:20:00           Cash     7.0           0.48  
10048       3.0  15/06/23  16:00:00    Credit card     4.0           0.48  
10049       2.0  25/02/21  12:25:00           Cash     7.0           0.48  
10050       3.0  26/09/20   9:48:00           Cash     3.0           0.33

In [7]:
df.describe()


invoice_id      quantity        rating  profit_margin
count  10051.000000  10020.000000  10051.000000   10051.000000
mean    5025.741220      2.353493      5.825659       0.393791
std     2901.174372      1.602658      1.763991       0.090669
min        1.000000      1.000000      3.000000       0.180000
25%     2513.500000      1.000000      4.000000       0.330000
50%     5026.000000      2.000000      6.000000       0.330000
75%     7538.500000      3.000000      7.000000       0.480000
max    10000.000000     10.000000     10.000000       0.570000

From above table we can already idetify anomalies like:

\> all **invoice_id, rating, profit_margin** has total **10051** number of rows while **quantity** has only **10020** rows only so it sugggest, we may identify some missing datas or NULL values

In [8]:
df.info

<bound method DataFrame.info of        invoice_id   Branch         City                category unit_price  \
0               1  WALM003  San Antonio       Health and beauty     $74.69   
1               2  WALM048    Harlingen  Electronic accessories     $15.28   
2               3  WALM067  Haltom City      Home and lifestyle     $46.33   
3               4  WALM064      Bedford       Health and beauty     $58.22   
4               5  WALM013       Irving       Sports and travel     $86.31   
...           ...      ...          ...                     ...        ...   
10046        9996  WALM056      Rowlett     Fashion accessories        $37   
10047        9997  WALM030   Richardson      Home and lifestyle        $58   
10048        9998  WALM050     Victoria     Fashion accessories        $52   
10049        9999  WALM032        Tyler      Home and lifestyle        $79   
10050       10000  WALM069     Rockwall     Fashion accessories        $62   

       quantity      date      

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10051 entries, 0 to 10050
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   invoice_id      10051 non-null  int64  
 1   Branch          10051 non-null  object 
 2   City            10051 non-null  object 
 3   category        10051 non-null  object 
 4   unit_price      10020 non-null  object 
 5   quantity        10020 non-null  float64
 6   date            10051 non-null  object 
 7   time            10051 non-null  object 
 8   payment_method  10051 non-null  object 
 9   rating          10051 non-null  float64
 10  profit_margin   10051 non-null  float64
dtypes: float64(3), int64(1), object(7)
memory usage: 863.9+ KB


From above table we can already idetify anomalies like:

* > All **columns or attribute** has total **10051** number of rows while **unit_price, quantity** has only **10020** rows only so it sugggest, we may identify some missing datas or NULL values
* > Data type of **unit_price** is  **_object_** but it must be of **_numeric_** type like **quantity**

**Now we will check for Duplicates**

In [10]:
# all duplicates
df.duplicated().sum()

np.int64(51)

Total _51_ duplictes

**Now we will count number of nulls**

In [11]:
df.isnull().sum()

invoice_id         0
Branch             0
City               0
category           0
unit_price        31
quantity          31
date               0
time               0
payment_method     0
rating             0
profit_margin      0
dtype: int64

So, _unit price and  quantity_ has total _31_ nulls  
**Now we will drop duplicates and again check number of nulls**

In [12]:
df.drop_duplicates(inplace=True)
df.duplicated().sum()

np.int64(0)

In [13]:
df.shape

(10000, 11)

Now we can observe the duplicates value is being removed and shape was previously (10051, 11) and now its (10000, 11)  
**Now lets drop the Null values**

In [14]:
#dropping all values with missing records
df.dropna(inplace=True)
df.isnull().sum()

invoice_id        0
Branch            0
City              0
category          0
unit_price        0
quantity          0
date              0
time              0
payment_method    0
rating            0
profit_margin     0
dtype: int64

In [15]:
df.shape

(9969, 11)

So clearly we can see the null count is 0 for each row, and the shape of the table is furthur modified from (10000,11) to (9969,11)

**Now we will take care of the data type of unit_price column**

The unit price column will be problamatic to deal with when we are performing any arithmatic calculations or funtions

In [16]:
#df['unit_price'].astype(float)

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[51], line 1
----> 1 df['unit_price'].astype(float)

File c:\Users\paula\Documents\My Workspace\Projects\Walmart_sales_dataAnalysis\.env_WalmartSales\Lib\site-packages\pandas\core\generic.py:6662, in NDFrame.astype(self, dtype, copy, errors)
   6656     results = [
   6657         ser.astype(dtype, copy=copy, errors=errors) for _, ser in self.items()
   6658     ]
   6660 else:
   6661     # else, only a single dtype is given
-> 6662     new_data = self._mgr.astype(dtype=dtype, copy=copy, errors=errors)
   6663     res = self._constructor_from_mgr(new_data, axes=new_data.axes)
   6664     return res.__finalize__(self, method="astype")

File c:\Users\paula\Documents\My Workspace\Projects\Walmart_sales_dataAnalysis\.env_WalmartSales\Lib\site-packages\pandas\core\internals\managers.py:430, in BaseBlockManager.astype(self, dtype, copy, errors)
    427 elif using_copy_on_write():
    428     copy = False
--> 430 return self.apply(
    431     "astype",
    432     dtype=dtype,
    433     copy=copy,
    434     errors=errors,
    435     using_cow=using_copy_on_write(),
    436 )
...
    132     # Explicit copy, or required since NumPy can't view from / to object.
--> 133     return arr.astype(dtype, copy=True)
    135 return arr.astype(dtype, copy=copy)

ValueError: could not convert string to float: '$74.69'
Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...

**_ValueError: could not convert string to float: '$74.69'_**  
Explicit copy, or required since NumPy can't view from / to object.

In [17]:
df['unit_price'] = df['unit_price'].str.replace('$', '').astype(float)

df.head()

invoice_id   Branch         City                category  unit_price  \
0           1  WALM003  San Antonio       Health and beauty       74.69   
1           2  WALM048    Harlingen  Electronic accessories       15.28   
2           3  WALM067  Haltom City      Home and lifestyle       46.33   
3           4  WALM064      Bedford       Health and beauty       58.22   
4           5  WALM013       Irving       Sports and travel       86.31   

   quantity      date      time payment_method  rating  profit_margin  
0       7.0  05/01/19  13:08:00        Ewallet     9.1           0.48  
1       5.0  08/03/19  10:29:00           Cash     9.6           0.48  
2       7.0  03/03/19  13:23:00    Credit card     7.4           0.33  
3       8.0  27/01/19  20:33:00        Ewallet     8.4           0.33  
4       7.0  08/02/19  10:37:00        Ewallet     5.3           0.48

Now the dollar sign is removed

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9969 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   invoice_id      9969 non-null   int64  
 1   Branch          9969 non-null   object 
 2   City            9969 non-null   object 
 3   category        9969 non-null   object 
 4   unit_price      9969 non-null   float64
 5   quantity        9969 non-null   float64
 6   date            9969 non-null   object 
 7   time            9969 non-null   object 
 8   payment_method  9969 non-null   object 
 9   rating          9969 non-null   float64
 10  profit_margin   9969 non-null   float64
dtypes: float64(4), int64(1), object(6)
memory usage: 934.6+ KB


data type of unit_price is converted successfully to float64

In [19]:
#check all columns
df.columns

Index(['invoice_id', 'Branch', 'City', 'category', 'unit_price', 'quantity',
       'date', 'time', 'payment_method', 'rating', 'profit_margin'],
      dtype='object')

In [20]:
df.columns = df.columns.str.lower()
df.columns

Index(['invoice_id', 'branch', 'city', 'category', 'unit_price', 'quantity',
       'date', 'time', 'payment_method', 'rating', 'profit_margin'],
      dtype='object')

In [21]:
df['total'] = df['unit_price'] * df['quantity']

df.head()

invoice_id   branch         city                category  unit_price  \
0           1  WALM003  San Antonio       Health and beauty       74.69   
1           2  WALM048    Harlingen  Electronic accessories       15.28   
2           3  WALM067  Haltom City      Home and lifestyle       46.33   
3           4  WALM064      Bedford       Health and beauty       58.22   
4           5  WALM013       Irving       Sports and travel       86.31   

   quantity      date      time payment_method  rating  profit_margin   total  
0       7.0  05/01/19  13:08:00        Ewallet     9.1           0.48  522.83  
1       5.0  08/03/19  10:29:00           Cash     9.6           0.48   76.40  
2       7.0  03/03/19  13:23:00    Credit card     7.4           0.33  324.31  
3       8.0  27/01/19  20:33:00        Ewallet     8.4           0.33  465.76  
4       7.0  08/02/19  10:37:00        Ewallet     5.3           0.48  604.17

In [22]:
# importing dependencies

# mysql
# host = localhost
# port = 3306
# user = root
# password = 'your_password'

In [23]:
df.to_csv('Walmart_clean_data.csv', index=False)

In [24]:
# import sqlite3
# conn = sqlite3.connect("WalmartDB.db")   # creates a new db file if not exists
# cursor = conn.cursor()

In [25]:
#my_sql connections
#"mysql+pymysql://user:password'@'localhost:3306/db_name"

eng_mysql = create_engine("mysql+pymysql://root@localhost:3307/walmart_db")
try:
    eng_mysql
    print("Connection Successed to mysql")
except:
    print("Unable to connect")

Connection Successed to mysql


In [26]:
df.to_sql(name='walmart', con=eng_mysql, if_exists='append', index=False)

9969